hudi-common/src/main/java/org/apache/hudi/common/model/SerializableIndexedRecord.java

```diff
@@ -0,0 +1,145 @@
+ * This allows the record to stay in the serialized form until the data needs to be accessed, which allows deserialization to be avoided if data is not read.
+ */
+class SerializableIndexedRecord implements GenericRecord, KryoSerializable {
+  private IndexedRecord record;
+  private byte[] recordBytes;
+
+  static SerializableIndexedRecord createInstance(IndexedRecord record) {
+    return record == null ? null : new SerializableIndexedRecord(record);
+  }
+
+  private SerializableIndexedRecord(IndexedRecord record) {
+    this.record = record;
+  }
+
+  @Override
+  public void put(int i, Object v) {
+    record.put(i, v);
+  }
+
+  @Override
+  public Object get(int i) {
+    return record.get(i);
+  }
+
+  @Override
+  public Schema getSchema() {
+    return record.getSchema();
+  }
+
+  private byte[] encodeRecord() {
+    if (recordBytes == null) {
+      recordBytes = HoodieAvroUtils.avroToBytes(record);
+    }
+    return recordBytes;
+  }
+
+  void decodeRecord(Schema schema) {
+    if (record == null) {
+      try {
+        record = HoodieAvroUtils.bytesToAvro(recordBytes, schema);
+      } catch (IOException e) {
+        throw new HoodieIOException("Failed to parse record into provided schema", e);
+      }
+    }
+  }
+
+  @Override
+  public void write(Kryo kryo, Output output) {
+    byte[] bytes = encodeRecord();
+    output.writeInt(bytes.length, true);
+    output.writeBytes(bytes);
+  }
+
+  @Override
+  public void read(Kryo kryo, Input input) {
+    int length = input.readInt(true);
+    this.recordBytes = input.readBytes(length);
+  }
+
+  @Override
+  public void put(String key, Object v) {
+    Schema.Field field = record.getSchema().getField(key);
+    record.put(field.pos(), v);
+  }
+
+  @Override
+  public Object get(String key) {
+    Schema.Field field = record.getSchema().getField(key);
+    return record.get(field.pos());
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null) {
+      return false;
+    }
+    if (o instanceof SerializableIndexedRecord) {
+      SerializableIndexedRecord that = (SerializableIndexedRecord) o;
+      ValidationUtils.checkArgument(record != null && that.record != null, "Records must be deserialized before equality check");
+      return record.equals(that.record);
+    } else if (o instanceof IndexedRecord) {
+      // If the other object is an IndexedRecord, we can compare it directly
+      IndexedRecord that = (IndexedRecord) o;
+      return record != null && record.equals(that);
+    }
+    return false;
+  }
+
+  IndexedRecord getRecord() {
+    ValidationUtils.checkArgument(record != null, "Record must be deserialized before accessing");
+    return record;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(record);
+  }
+
+  @Override
+  public String toString() {
+    return record == null ? "SERIALIZED: " + new String(recordBytes) : record.toString();
```

Comment:
we probably have a util to hex print byte arrays? printing record is a bit sensitive. please check once that we are not accidentally logging this more than before etc. Resolve if you have done it already